src/DotVVM.Framework/Compilation/Static/StaticViewCompiler.cs

```diff
@@ -0,0 +1,277 @@
+                file.LastWriteDateTimeUtc);
+            var className = DefaultControlBuilderFactory.GetClassFromFileName(file.FileName) + "ControlBuilder";
+            string fullClassName = namespaceName + "." + className;
+            var refObjectSerializer = configuration.ServiceProvider.GetRequiredService<RefObjectSerializer>();
+            var emitter = new DefaultViewCompilerCodeEmitter();
+            var assemblyCache = configuration.ServiceProvider.GetRequiredService<CompiledAssemblyCache>();
+            var bindingCompiler = configuration.ServiceProvider.GetRequiredService<IBindingCompiler>();
+            var compilingVisitor = new ViewCompilingVisitor(emitter, assemblyCache, bindingCompiler, className);
+            resolvedView.Accept(compilingVisitor);
+
+            // HACK: In 2.4.0, ReflectionUtils.FindType expects to find the controls in the current assembly,
+            //       which is bollocks.
+            if (resolvedView.Metadata.Type != typeof(DotvvmView))
+            {
+                emitter.ResultControlTypeSyntax = emitter.ParseTypeName(resolvedView.Metadata.Type);
+            }
+
+            if (resolvedView.Directives.ContainsKey("masterPage"))
+            {
+                // make sure that the masterpage chain is already compiled
+                _ = GetView(resolvedView.Directives["masterPage"].Single().Value);
+            }
+
+            var syntaxTree = emitter.BuildTree(namespaceName, className, viewPath).Single();
+            var references = emitter.UsedAssemblies.Select(a =>
+                MetadataReference.CreateFromFile(a.Key.Location).WithAliases(new[] { a.Value, "global" }));
+
+            var compilation = CSharpCompilation.Create(
+                assemblyName: $"{fullClassName}.Compiled",
+                syntaxTrees: new[] { syntaxTree },
+                references: baseReferences.Concat(references),
+                options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+            using var memoryStream = new MemoryStream();
+            var result = compilation.Emit(memoryStream);
+            if (!result.Success)
+            {
+                reports.Add(new Report(viewPath, -1, -1, "Compilation failed. This is likely a bug in the DotVVM compiler."));
+                return view.WithReports(reports);
+            }
+
+            var assembly = Assembly.Load(memoryStream.ToArray());
+            return view.WithAssembly(assembly)
+                .WithViewType(resolvedView.Metadata.Type)
+                .WithDataContextType(resolvedView.DataContextTypeStack.DataContextType);
+        }
+
+        /// <summary>
+        /// HACK: Because as of 2.4.0, DotVVM gets a list of all assemblies only from the Default DependencyContext,
+        ///       a problem arises, because in this Compiler, DotVVM itself isn't in the Default DependencyContext, thus
+        ///       I need to invoke the static constructors of controls myself.
+        /// </summary>
```

Comment:
You claim to be invoking constructors on the Dotvvm.Framework assembly, but it actually runs though many more assemblies, right?

Reply:
I removed this particular hack in favor of a better one (or a much worse one, depending on your point of view) in `DependencyContextCompilerExecutor`.